src/components/VerificationCode/index.tsx

```diff
@@ -65,99 +103,108 @@ const VerificationCode = ({
     }
   }
 
-  const inputOnChange = e => {
-    const index = parseInt(e.target.dataset.id, 10)
+  const inputOnChange = (event: ChangeEvent<HTMLInputElement>) => {
+    const index = parseInt(String(event.target.dataset.id), 10)
+    let { value } = event.target
     if (type === 'number') {
-      e.target.value = e.target.value.replace(/[^\d]/gi, '')
+      value = event.target.value.replace(/[^\d]/gi, '')
     }
     const newValues = [...values]
 
     if (
-      e.target.value === '' ||
-      (type === 'number' && !e.target.validity.valid)
+      value === '' ||
+      (type === 'number' && !new RegExp(event.target.pattern).test(value))
     ) {
       newValues[index] = ''
       setValues(newValues)
 
       return
     }
 
-    const { value } = e.target
     const sanitizedValue = value[0] // in case more than 1 char, we just take the first one
     newValues[index] = sanitizedValue
     setValues(newValues)
     const nextIndex = Math.min(index + 1, fields - 1)
     const next = inputRefs[nextIndex]
 
-    if (next) {
-      next.current.focus()
-    }
+    next?.current?.focus()
 
     triggerChange(newValues)
   }
 
-  const inputOnKeyDown = e => {
-    const index = parseInt(e.target.dataset.id, 10)
+  const inputOnKeyDown: KeyboardEventHandler<HTMLInputElement> = (
+    event: KeyboardEvent,
+  ) => {
+    if (!event || !(event.target instanceof HTMLInputElement)) {
+      return
+    }
+
+    const index = parseInt(event.target.dataset.id as string, 10)
     const prevIndex = index - 1
     const nextIndex = index + 1
     const first = inputRefs[0]
     const last = inputRefs[inputRefs.length - 1]
     const prev = inputRefs[prevIndex]
     const next = inputRefs[nextIndex]
     const vals = [...values]
-    switch (e.keyCode) {
+
+    switch (event.keyCode) {
       case KEY_CODE.backspace:
-        e.preventDefault()
+        event.preventDefault()
         if (values[index]) {
           vals[index] = ''
           setValues(vals)
           triggerChange(vals)
         } else if (prev) {
           vals[prevIndex] = ''
-          prev.current.focus()
+          prev.current?.focus()
           setValues(vals)
           triggerChange(vals)
         }
         break
       case KEY_CODE.left:
-        e.preventDefault()
+        event.preventDefault()
         if (prev) {
-          prev.current.focus()
+          prev.current?.focus()
         }
         break
       case KEY_CODE.right:
-        e.preventDefault()
+        event.preventDefault()
         if (next) {
-          next.current.focus()
+          next.current?.focus()
         }
         break
       case KEY_CODE.up:
-        e.preventDefault()
+        event.preventDefault()
         if (first) {
-          first.current.focus()
+          first.current?.focus()
         }
         break
       case KEY_CODE.down:
-        e.preventDefault()
+        event.preventDefault()
         if (last) {
```

Comment:
Why check last ? It will be always defined. Instead last.current can be undefined